File: Substring with Concatenation of All Words_hard.py
from collections import Counter

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        n, m, num = len(s), len(words), len(words[0])
        wordmap = Counter(words)
        ans = []

        for i in range(n - m * num + 1):
            hashmap = {}
            j = 0
            while j < m:
                key = s[i + j * num: i + j * num + num]
                if key in wordmap:
                    hashmap[key] = hashmap[key] + 1 if key in hashmap else 1
                    if hashmap[key] > wordmap[key]:
                        break
                else:
                    break
                j = j + 1
            if j >= m:
                ans.append(i)
        return ans

File: Substring with Concatenation of All Words_hard.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        int n = s.size(), m = words.size(), len = words[0].size(), i, j;
        unordered_map<string,int> wordmap = {}, dict={};
        vector<int> ans = {};
        
        for(auto str:words)
            wordmap[str]++;
        for(i = 0; i <= n - m * len; i++){
            dict = {};
            for(j = 0; j <= m; j++){
                string aux = s.substr(i + j * len, len);
                if(wordmap.count(aux)){
                    dict[aux]++;
                    if(dict[aux] > wordmap[aux])
                        break;
                }
                else
                    break;
            }
            if (j >= m)
                ans.push_back(i);
        }
        return ans;
    }
};